servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleTerminalSignalConsumer.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.Cancellable;
+import io.servicetalk.concurrent.SingleSource;
+
+import javax.annotation.Nullable;
+
+/**
+ * A contract that provides discrete callbacks for various ways in which a {@link SingleSource.Subscriber} can
+ * terminate.
+ *
+ * @param <T> Type of the result of the {@link Single}.
+ */
+public interface SingleTerminalSignalConsumer<T> {
```

Comment:
not strong on internal type vs top level type, but leaning toward external type so it is more consistent with `TerminalSignalConsumer` and we can expose `RunnableSingleTerminalSignalConsumer` in the future if necessary without cluttering `Single`'s public exposure.